### PR TITLE
Fix unsupported type detection when serializing polymorphic System.Type instances

### LINKED ISSUE #67588: System.Text.Json.JsonSerializer does not work with Dictionary<string, object>
### Description

When a `Dictionary<string, object>` needs serialization, the .NET default `System.Text.Json.JsonSerializer` throws an exception:

> System.InvalidOperationException: Method may only be called on a Type for which Type.IsGenericParameter is true.

I have provided a simple piece of code example to reproduce it. I tried with `Newtonsoft.Json.JsonConvert` and it works flawlessly.

### Reproduction Steps

```
 public void TestDictionarySerialize()
        {
            var dictionary = new Dictionary<string, object> {
                {"Hello", typeof(string)}, 
                {"World", typeof(string)}
            };

            //This serialization fails
            System.Text.Json.JsonSerializer.Serialize(dictionary,
                typeof(Dictionary<string, object>));

           //This serialization fails
            System.Text.Json.JsonSerializer.Serialize(dictionary);

            //This serialization succeeds
            Newtonsoft.Json.JsonConvert.SerializeObject(dictionary);
        }
```

### Expected behavior

`System.Text.Json.JsonSerializer` should not throw any Exception

### Actual behavior

`System.Text.Json.JsonSerializer` throws:

> System.InvalidOperationException: Method may only be called on a Type for which Type.IsGenericParameter is true.

### Regression?

_No response_

### Known Workarounds

`Newtonsoft.Json.JsonConvert` works flawlessly

### Configuration

- .NET 6.0.200
- Windows 10 Enterprise 21H2 19044.1586
- x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #67588.

### CODE REVIEW

**Review by krwq — APPROVED** [review]
LGTM with suggestion

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/tests/Common/UnsupportedTypesTests.cs [hunk 1]
```diff
@@ -72,6 +72,7 @@ async Task RunTest<T>(string json)
         [Fact]
         public async Task SerializeUnsupportedType()
         {
+            // TODO refactor to Xunit theory
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/UnsupportedTypesTests.cs:75]: It's a pervasive pattern so I'd rather it be done in a separate PR.

File: src/libraries/System.Text.Json/tests/Common/UnsupportedTypesTests.cs [hunk 2]
```diff
@@ -124,6 +124,20 @@ async Task RunTest<T>(T value)
                     serialized = await Serializer.SerializeWrapper(obj, new JsonSerializerOptions { IgnoreNullValues = true });
                     Assert.Equal(@"{}", serialized);
                 }
+
+#if !BUILDING_SOURCE_GENERATOR_TESTS
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Common/UnsupportedTypesTests.cs:128]: Why isn't this being run in source-gen mode?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/Common/UnsupportedTypesTests.cs:128]: Polymorphic scenaria will throw `InvalidOperationException` before we get the chance to exercise the changed codepath.

Comment (layomia) [reply src/libraries/System.Text.Json/tests/Common/UnsupportedTypesTests.cs:128]: Got it. FWIW when there are consequential divergences in source-gen/reflection behavior, we add a separate test variant for source-gen, e.g. [here](https://github.com/dotnet/runtime/blob/424a09cb81c678fb1ba1c27211b80aba2de070ad/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/Serialization/PropertyVisibilityTests.cs#L28).

**Review by layomia — APPROVED** [review]

**Comment by eiriktsarpalis** [comment]
Test failures unrelated to Json

**Pull request merged**